File: Baitap2/Complete/Blackbox/MergeSort.java
import java.util.Arrays;
//import java.util.Scanner;

public class MergeSort {
    /*public static void main(String[] args) {
        Scanner scanner = new Scanner(System.in);
        int n = scanner.nextInt();
        int[] arr = new int[n];
        for(int i = 0; i < n; i ++)
            arr[i] = scanner.nextInt();
        
        int[] result = sort(arr);
        for(int x : result)
            System.out.print(x + " ");
        scanner.close();
    }*/

    public static int[] merge(int[] left, int[] right) {
        // Corner cases: either left or right equals null
        if (left == null)
            return right;
        if (right == null)
            return left;

        int[] result = new int[left.length + right.length];
        int idLeft = 0, idRight = 0;
        for(int i = 0; i < result.length; i ++) {
            if (idLeft == left.length)
                result[i] = right[idRight ++];
            else if (idRight == right.length)
                result[i] = left[idLeft ++];
            else {
                if (left[idLeft] < right[idRight])
                    result[i] = left[idLeft ++];
                else
                    result[i] = right[idRight ++];
            }
        }

        return result;
    }

    public int[] sort(int[] arr) {
        // Corner cases: Either array equals to null, or the array has length less than or equal to 1
        if (arr == null || arr.length <= 1)
            return arr;
        int middle = arr.length / 2;
        int[] left = sort(Arrays.copyOfRange(arr, 0, middle));
        int[] right = sort(Arrays.copyOfRange(arr, middle, arr.length));
        int[] result = merge(left, right);
        return result;
    }
}

File: Baitap2/Complete/Blackbox/MergeSort.py
def merge(L1, L2):      
        if (len(L1) == 0): return L2
        if (len(L2) == 0): return L1
        i = i1 = i2 = 0
        L = []
        while(i1 < len(L1) and i2 < len(L2)):
                if (L1[i1] < L2[i2]): 
                        L.append(L1[i1])
                        i1 += 1
                else:
                        L.append(L2[i2])
                        i2 += 1
        
        while(i1 < len(L1)):
                L.append(L1[i1])
                i1 += 1
        while(i2 < len(L2)):
                L.append(L2[i2])
                i2 += 1
        return L

def mergeSort(ent):
        if len(ent) > 1:
                m = len(ent) // 2
                L1 = ent[:m]
                L2 = ent[m:len(ent)]
                L = merge(mergeSort(L1) , mergeSort(L2))
                return L
        else: return ent
        
arr = [5, 6, -6, 1, 3, -7]
print(arr)

print(mergeSort(arr))